### PR TITLE
bugfix: 'spack compiler add'  properly resolves relative path to absolute path

### LINKED ISSUE #11782: spack compiler add uses relative path
If fed a relative path, `spack compiler add` will write the relative path to compilers.yaml rather than dereference it to the absolute path. This behavior began with commit 6d56d4545472ddff1cdc0e8fd49d8260428a141e according to `git bisect`.

### Steps to reproduce the issue

```console
$ spack compiler add --scope=site ../apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/
==> Added 1 new compiler to /opt/packages/gpjohnsn/opt/ssoft/spack/etc/spack/compilers.yaml
    intel@19.0.3.199
==> Compilers are defined in the following files:
    /opt/packages/gpjohnsn/opt/ssoft/spack/etc/spack/compilers.yaml
$ cat etc/spack/compilers
...
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos7
    paths:
      cc: ../apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/icc
      cxx: ../apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/icpc
      f77: ../apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/ifort
      fc: ../apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/ifort
    spec: intel@19.0.3.199
    target: x86_64
...
```
```console
Previous, and expected, output:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos7
    paths:
      cc: /opt/packages/gpjohnsn/opt/ssoft/apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/icc
      cxx: /opt/packages/gpjohnsn/opt/ssoft/apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/icpc
      f77: /opt/packages/gpjohnsn/opt/ssoft/apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/ifort
      fc: /opt/packages/gpjohnsn/opt/ssoft/apps/linux-centos7-x86_64/gcc-4.8.5/intel-19.0.3-4vf3wugrvromolotwrco5vuzydnts3bv/bin/ifort
    spec: intel@19.0.3.199
    target: x86_64
...

```
Loading the respective module first will do the right thing but using a relative path in the `spack compiler` command used to work. Also, it seems that a relative path should never be written to compilers.yaml.

### PR DESCRIPTION
**Pull request by carsonwoods**
Fixes #11782. Spack was not properly resolving relative paths to absolute paths when a relative path was passed to `spack compiler add [PATH]`. Now, if provided a relative path, the absolute path is written to compilers.yaml rather than relative path.

### CODE REVIEW

File: lib/spack/spack/architecture.py [hunk 1]
```diff
@@ -226,7 +226,7 @@ def __repr__(self):
         return self.__str__()
 
     def _cmp_key(self):
-        return self.name, self.version
+        return (self.name, self.version)
```
Comment (adamjstewart) [inline lib/spack/spack/architecture.py]: Was this change accidental?

Comment (carsonwoods) [reply lib/spack/spack/architecture.py]: Whoops yes it was. I will fix it.

Comment (carsonwoods) [reply lib/spack/spack/architecture.py]: Ok, its fixed now.

File: lib/spack/llnl/util/filesystem.py [hunk 2]
```diff
@@ -1415,6 +1415,6 @@ def search_paths_for_executables(*path_hints):
 
         bin_dir = os.path.join(path, 'bin')
         if os.path.isdir(bin_dir):
-            executable_paths.append(bin_dir)
+            executable_paths.append(os.path.abspath(bin_dir))
```
Comment (adamjstewart) [inline lib/spack/llnl/util/filesystem.py]: This works if the user gives a `PATH` like `../gcc/` but not if the user gives a `PATH` like `../gcc/bin`. I think this logic should be moved farther up.

Comment (carsonwoods) [reply lib/spack/llnl/util/filesystem.py]: Oh good call. I have pushed a change that fixes it.

File: lib/spack/llnl/util/filesystem.py [hunk 3]
```diff
@@ -1411,7 +1411,7 @@ def search_paths_for_executables(*path_hints):
         if not os.path.isdir(path):
             continue
 
-        executable_paths.append(path)
+        executable_paths.append(os.path.abspath(path))
```
Comment (adamjstewart) [inline lib/spack/llnl/util/filesystem.py]: Now `../gcc/bin` will work, but `../gcc/` won't. You should use:
```python
path = os.path.abspath(path)
```

Comment (carsonwoods) [reply lib/spack/llnl/util/filesystem.py]: Ok, its fixed now.

**Comment by scheibelp** [comment]
@carsonwoods thanks for the fix!
@adamjstewart thanks for the review!

**Pull request merged**